codec/src/codec.rs

```diff
@@ -40,25 +40,35 @@ pub trait Codec: Sized {
 }
 
 /// Trait for types that have a fixed-length encoding
-pub trait SizedCodec<const N: usize>: Codec {
+pub trait SizedCodec: Codec {
+    /// The encoded length of this value.
+    const LEN_ENCODED: usize;
+
     /// Returns the encoded length of this value.
     ///
     /// Should not be overridden by implementations.
     fn len_encoded(&self) -> usize {
-        N
+        Self::LEN_ENCODED
     }
 
     /// Encodes a value to fixed-size bytes.
-    fn encode_fixed(&self) -> [u8; N] {
+    fn encode_fixed<const N: usize>(&self) -> [u8; N] {
+        // Ideally this is a compile-time check, but we can't do that in the current Rust version
+        // without adding a new generic parameter to the trait.
+        assert_eq!(
+            N,
+            Self::LEN_ENCODED,
+            "Can't encode {} bytes into {} bytes",
+            Self::LEN_ENCODED,
+            N
+        );
+
         self.encode().try_into().unwrap()
     }
 }
 
 /// Trait for codec read operations
 pub trait Reader {
-    /// Reads a value of type T
-    fn read<T: Codec>(&mut self) -> Result<T, Error>;
-
```

Comment:
Allowed a client to easily accidentally write infinite recursion code that ends up with a stack overflow:
```
impl Codec for X {
    fn read(reader: &mut impl Reader) -> Result<Self, Error> {
        reader.read(self)
    }
}
```

instead, force use of `<Y>::read(reader);`